### PR TITLE
new preprocessing and training for structured data

### PR DESCRIPTION
**Pull request by brandondutra**
prediction is not done

preprocessing:
--local uses pandas
--cloud: if given bq table, does analysis on that. If given csv files, uses federated tables.

training:
--starts from raw csv data
--all transformations are done in training. 
--transform types depend on if linear/dnn model is used.

### CODE REVIEW

**Comment by brandondutra** [comment]
schema file example
[
    {
        "mode": "REQUIRED",
        "name": "key",
        "type": "STRING"
    },
    {
        "mode": "NULLABLE",
        "name": "target",
        "type": "INTEGER"
    },
    {
        "mode": "NULLABLE",
        "name": "num1",
        "type": "FLOAT"
    },
    {
        "mode": "NULLABLE",
        "name": "num2",
        "type": "INTEGER"
    },
    {
        "mode": "NULLABLE",
        "name": "num3",
        "type": "FLOAT"
    },
    {
        "mode": "NULLABLE",
        "name": "str1",
        "type": "STRING"
    },
    {
        "mode": "NULLABLE",
        "name": "str2",
        "type": "STRING"
    },
    {
        "mode": "NULLABLE",
        "name": "str3",
        "type": "STRING"
    }  
]

**Comment by brandondutra** [comment]
feature type example (consumed by preprocess). Default and type are required
{
  "key": {"default": -1, "type": "key"},
  "target": {"default": "unknown", "type": "categorical"},
  "num1": {"default": 0.0, "type": "numerical"},
  "num2": {"default": 0, "type": "numerical"},
  "num3": {"default": 0.0, "type": "numerical"},
  "str1": {"default": "black", "type": "categorical"},
  "str2": {"default": "abc", "type": "categorical"},
  "str3": {"default": "car", "type": "categorical"}
}

**Comment by brandondutra** [comment]
example transforms file
{
  "num1": {"transform": "scale"},
  "num2": {"transform": "max_abs_scale", "value": 4},
  "str1": {"transform": "one_hot"},
  "str2": {"transform": "embedding", "embedding_dim": 3},
  "target": {"transform": "target"}
 }

supported numerical transforms: scale, max_abs_scale, identity (default)
supported categorical transforms for dnn models: hash_embedding, embedding, hash_one_hot, one_hot (default)
supported categorical transforms for linear models: hash_sparse, sparse (default)

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+import json
+
+import google.cloud.ml as ml
+
+
+INPUT_FEATURES_FILE = 'input_features.json'
+SCHEMA_FILE = 'schema.json'
+
+NUMERICAL_ANALYSIS = 'numerical_analysis.json'
+CATEGORICAL_ANALYSIS = 'vocab_%s.csv'
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py]: These definitions are repeated in util,py, local_preprocess.py. Can we centralize them to util.py?

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 2]
```diff
@@ -91,6 +105,26 @@ def end(self, session=None):
...
 
+# ==============================================================================
+# Reading the input csv files and parsing its output into tensors.
+# ==============================================================================
+
+def parse_example_tensor(examples, train_config):
+
+  # record_defaults are used by tf.decode_csv to insert defaults, and to infer
+  # the datatype. 
+  record_defaults = [[train_config['csv_defaults'][name]]
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: csv_defaults can be optional too? We can use default value based on schema. We can require no missing values if defaults are not provided?

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 3]
```diff
@@ -115,9 +149,12 @@ def read_examples(input_files, batch_size, shuffle, num_epochs=None):
...
   filename_queue = tf.train.string_input_producer(files, num_epochs, shuffle)
 
-  options = tf.python_io.TFRecordOptions(
-      compression_type=tf.python_io.TFRecordCompressionType.GZIP)
-  example_id, encoded_example = tf.TFRecordReader(options=options).read_up_to(
+  #options = tf.python_io.TFRecordOptions(
+  #    compression_type=tf.python_io.TFRecordCompressionType.GZIP)
+  #example_id, encoded_example = tf.TFRecordReader(options=options).read_up_to(
+  #    filename_queue, batch_size)
+
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: remove?

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 4]
```diff
@@ -140,67 +177,180 @@ def read_examples(input_files, batch_size, shuffle, num_epochs=None):
...
+        transform_config = train_config['transforms'].get(name, {})
+        transform_name = transform_config.get('transform', None)
+        if transform_name == 'scale':
+          features[name] = _scale_tensor(
+              features[name], 
+              range_min=numerical_anlysis[name]['min'], 
+              range_max=numerical_anlysis[name]['max'], 
+              scale_min=-1, 
+              scale_max=1)
+        elif transform_name == 'max_abs_scale':
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: I wonder if we can combine scale and max_abs_scale, and make scale value default to 1?

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 5]
```diff
@@ -214,77 +364,236 @@ def _scale_tensor(tensor, range_min, range_max, scale_min, scale_max):
...
-  this function is given to tf.learn to further process. This function extracts
-  the ids tensors from ml.features.FeatureMetadata.parse_features and throws
-  away the values tensor.
+def get_vocabulary(preprocess_output_dir, name):
+  vocab_file = os.path.join(preprocess_output_dir, CATEGORICAL_ANALYSIS % name)
+  if not ml.util._file.file_exists(vocab_file):
+    raise ValueError('File %s not found in %s' % (CATEGORICAL_ANALYSIS % name, preprocess_output_dir))
+
+  df = pd.read_csv(StringIO(ml.util._file.load_file(vocab_file)),
+                   header=0)
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: header=0 means using the first line as header. It seems in local_preprocess, you have "df.to_csv(None, header=False)" so no header?

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 6]
```diff
@@ -214,77 +364,236 @@ def _scale_tensor(tensor, range_min, range_max, scale_min, scale_max):
...
+  index_values = df['index'].values.tolist()
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: Perhaps we can just sort the labels so we don't need the index.

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py (hunk 1 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py]: Cannot do this if the package also supports the command line interface and the import package interface. These constants are in local/cloud_preprocess.py and util.py. Once I have a better testing story, I can merge all constants to a file

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py (hunk 6 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: removed the index.

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py (hunk 5 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: Update so that the vocab files are in the form:
label1
label2
label3EOF

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py (hunk 4 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: good idea.

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py (hunk 3 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py (hunk 2 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: I have to have a default value when reading data from csv in TF, so there is no way of knowing at training time if data is missing. Picking a bad default value would make a model useless, so I would want to just ask the user for these values.

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/task.py [hunk 7]
```diff
@@ -170,127 +154,43 @@ def get_export_signature(examples, features, predictions):
...
-    # Load the metadata file, transforms file, and schema file.
-    metadata = ml.features.FeatureMetadata.get_metadata(args.metadata_path)
-    transform_config = json.loads(ml.util._file.load_file(args.transforms_file))
-    schema_config = json.loads(ml.util._file.load_file(args.schema_file))
+    # Merge schema, input features, and transforms.
+    train_config = util.merge_metadata(args.preprocess_output_dir, args.transforms_file)
+
+    #metadata = ml.features.FeatureMetadata.get_metadata(args.metadata_path)
+    #transform_config = json.loads(ml.util._file.load_file(args.transforms_file))
+    #schema_config = json.loads(ml.util._file.load_file(args.schema_file))
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/task.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/local_preprocess.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    for name, config in feature_types.iteritems():
+      # Update numerical analsysis
+      if config['type'] == 'numerical':
+        numerical_results[name]['min'] = min(numerical_results[name]['min'], 
+                                             df[name].min())
+        numerical_results[name]['max'] = max(numerical_results[name]['max'],
+                                             df[name].max())
+      elif config['type'] == 'categorical':
+        # Update categorical analsysis
+        categorical_results[name].update(set(df[name].unique()))
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/local_preprocess.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/local_preprocess.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    if config['type'] == 'numerical':
+      numerical_results[name] = {'min': float('inf'), 'max': float('-inf')}
+
+  # initialize categorical_results
+  categorical_results = collections.defaultdict(set)
+
+  # for each file, update the min/max values from that file, and update the set
+  # of unique labels.
+  for input_file in input_files:
+    df = pd.read_csv(StringIO(ml.util._file.load_file(input_file)),
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/local_preprocess.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+    sys.stdout.write('Running categorical analysis...')
+    for name in categorical_columns:
+      if args.bigquery_table:
+        table_name = parse_table_name(args.bigquery_table)
+      else:
+        table_name = 'table_name'
+
+      # BQ does not have a distinct function, or a row number that starts at 0.
+      sql = """
+          SELECT
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+      raise ValueError('--input_file_pattern must point to files on GCS')
+
+  return args
+
+def parse_table_name(bigquery_table):
+  """Giving a string a:b.c, returns 'b.c'"""
+
+  id_name = bigquery_table.split(':')
+  if len(id_name) != 2:
+    raise ValueError('Bigquery table name should be in the form '
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py]: It is, cloud_preprocess in _package pulls in the default if missing.

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+
+  args = parser.parse_args(args=argv[1:])
+  print(args)
+
+  if not args.output_dir.startswith('gs://'):
+    raise ValueError('--output_dir must point to a location on GCS')
+
+  if args.bigquery_table:
+    if args.schema_file or args.input_file_pattern :
+      raise ValueError('If using --bigquery_table, then --schema_file and '
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py]: added a todo

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+  """
+  parser = argparse.ArgumentParser(
+      description='Runs Preprocessing on structured data.')
+  parser.add_argument('--output_dir',
+                      type=str,
+                      required=True,
+                      help='Google Cloud Storage which to place outputs.')
+  parser.add_argument('--input_feature_types',
+                      type=str,
+                      required=True,
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py]: I'll hold off on making things optional for now. Feature types also encodes the default values for training. Picking default values for training is not something I want to do, especially for categorical columns. I want to revisit this late.

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+import sys
+import json
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/cloud_preprocess.py]: done.

File: solutionbox/structured_data/datalab_solutions/structured_data/preprocess/local_preprocess.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+  args = parser.parse_args(args=argv[1:])
+
+  # Make sure the output folder exists if local folder.
+  file_io.recursive_create_dir(args.output_dir)
+
+  return args
+
+
+def run_numerical_categorical_analysis(args, feature_types, schema_list):
+  """Makes the nuermical and categorical analysis files.
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/preprocess/local_preprocess.py]: typo

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 16]
```diff
@@ -140,67 +172,173 @@ def read_examples(input_files, batch_size, shuffle, num_epochs=None):
...
-            bucket_size=N+1)
-      feature_columns.append(tf.contrib.layers.embedding_column(sparse_column, 
-                                                                dim))
-      # TODO(brandon): check model type is dnn.
-    else:
-      print('ERROR: unkown categorical transform name %s in %s' % 
-          (name, str(transform_type)))
-      sys.exit(1)
+  # Do target transform
+  with tf.name_scope('categorical_feature_preprocess') as scope:
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: Should be target_feature_preprocess?
